src/hotspot/share/oops/method.cpp

```diff
@@ -2254,10 +2254,12 @@ bool Method::is_method_id(jmethodID mid) {
 Method* Method::checked_resolve_jmethod_id(jmethodID mid) {
   if (mid == NULL) return NULL;
   Method* o = resolve_jmethod_id(mid);
-  if (o == NULL || o == JNIMethodBlock::_free_method || !((Metadata*)o)->is_method()) {
+  if (o == NULL || o == JNIMethodBlock::_free_method) {
     return NULL;
   }
-  return o;
+  // Method should otherwise be valid. Assert for testing.
+  assert(is_valid_method(o), "should be valid jmethodid");
+  return o->method_holder()->is_loader_alive() ? o : NULL;
```

Comment:
Let me ask the question a different way. Here's the two lines of code:
```
assert(is_valid_method(o), "should be valid jmethodid");
return o->method_holder()->is_loader_alive() ? o : NULL;
```

If `is_loader_alive()` can return false at this point, then what does
the previous `is_valid_method(o)` call return in that case? Should the
`assert()` call only be made in the case where `is_loader_alive()` returns
true?

Based on what you said in your reply, the normal expectation is that the
`assert()` call will pass because at that point we must have a valid Method.
(I'm ignoring corruption here). If we have a valid Method, then that implies
that the `is_loader_alive()` call must also return true. So if we're expecting
the Method to be valid at the point of the `assert()` call, then why are we
checking `is_loader_alive()`? It could be for robustness, but in that case,
I would add a comment making that clear.

Reply:
The assert is to check for corruptness but even if the loader isn't alive, the Method* should be valid.  For concurrent unloading (ie. zgc), checking that the method loader is alive goes like this:
method->method_holder() points to the method's InstanceKlass but the ->is_loader_alive will go to the ClassLoaderData and do a weak load of the holder oop (class_loader or mirror oop).  If the weak load fails, the loader is no longer alive, but the concurrent collector hasn't marked it as unloaded yet, but it will.  After the safepoint, the code is then safe to purge and that will reclaim memory for the Method and it's holder.

// Unloading support
bool ClassLoaderData::is_alive() const {
  bool alive = keep_alive()         // null class loader and incomplete non-strong hidden class.
      || (_holder.peek() != NULL);  // and not cleaned by the GC weak handle processing.

  return alive;
}